### ISSUE TITLE
Severe Performance Degradation for S3 using SDK 1.8.32

### ISSUE THREAD

**Issue by lintonv** [aws/aws-sdk-cpp#1440]
Confirm by changing [ ] to [x] below:
- [X] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/welcome.html) and [API reference](http://sdk.amazonaws.com/cpp/api/LATEST/index.html)
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-cpp/issues) and didn't find any solution

**Platform/OS/Hardware/Device**
What are you running the sdk on?
Ubuntu 18.04 (x86_64)

**Describe the question**
We have been using SDK version 1.7.336 for a while. But we need some of the new HttpClient features in 1.8. So we upgraded to 1.8.32. 

We saw our S3 transactions - puts, gets, lists all decrease in performance significantly 

For example - for uploading data to S3 - 
```
SDK 1.7.336, Speed - 1.5G to 1.75G
SDK 1.8.32, Speed - 50M to 150M 
```

**This is a significant performance regression** 

Here's the build info (we use bazel) 
```
aws_copts = [
    "-DPLATFORM_LINUX",
    "-DAWS_SDK_VERSION_MAJOR=1",
    "-DAWS_SDK_VERSION_MINOR=8",
    "-DAWS_SDK_VERSION_PATCH=32",
    "-DENABLE_CURL_CLIENT",
    "-DENABLE_OPENSSL_ENCRYPTION",
]
```
How do I track where the problem is? 

**Logs/output**
Nothing from a logs point of view

**Comment by KaibaLopez** [comment]
Hi @lintonv ,
That's interesting, would it be possible for you to share the code you're using to test these?

**Comment by lintonv** [comment]
Hi @KaibaLopez 

This is the gist of the code - 

```
void S3Storage::upload_file(const std::string &bucket_name,
                            const std::string &object_key,
                            const std::string &file_path) const
{

    Aws::S3::Model::PutObjectRequest object_request;

    object_request.WithBucket(to_aws_str(bucket_name)).WithKey(to_aws_str(object_key));

    // The allocation tag is used for memory tracking purposes only - this char* could be anything
    // See Aws::MakeShared usage for details.
    constexpr auto mem_alloc_tag = "fstream_sptr";
    
    auto input_data = Aws::MakeShared<Aws::Utils::FStreamWithFileName>(
        mem_alloc_tag, to_aws_str(file_path), std::ios_base::in | std::ios_base::binary);
    
    object_request.SetBody(input_data);
    
    auto tag = set_upload_request_properties(object_request, object_key);

    auto path    = fmt::format("s3://{}/{}", bucket_name, object_key);
    // retry logic
    auto outcome = with_retry([this, &object_request]() { return this->s3_client_->PutObject(object_request); }, path);

    if (!outcome.IsSuccess())
    {
        // debug stmt
    }

    try
    {
        add_tag_to_object(bucket_name, object_key, tag.first, tag.second);
    }
    catch (S3Exception &e)
    {
        //debug stmt
    }
}
```

This is the same code we use for SDK Version 1.7.336 (and the latest version 1.8.32)

**Comment by lintonv** [comment]
Thank you @KaibaLopez. Was this documented somewhere? Did I just miss that? If it is not documented, could you please make sure it is?

**Comment by lintonv** [comment]
@KaibaLopez - we saw a segfaults and other issues with that flag on EC2 environments. We have a system that has a mixtures of code running on EC2 and running outside AWS but depending on S3 clients. 

Is there an alternative way to not use that flag? Maybe configure the client in a different way?

**Comment by KaibaLopez** [comment]
@lintonv ,
Sorry we are working on a better way to do it but currently the only way to avoid these involve environment variables. If the AWS_EC2_METADATA_DISABLED is problematic for you, you could also set AWS_DEFAULT_REGION or AWS_REGION to your region...
Though honestly if having the env variable as true causes troubles later on, the best workaround I can think of is having something like:
```
setenv("AWS_EC2_METADATA_DISABLED", "true", 1);
Aws::Client::ClientConfiguration clientConfig("default"); // use this configuration for all aws clients or do the same for other client configurations
setenv("AWS_EC2_METADATA_DISABLED", "false", 1);
```
either that or changing the source code directly-  on clientConfiguration.cpp remove lines 127-134. 
Hope one of these works for you.

**Comment by lintonv** [comment]
Thanks @KaibaLopez 
What will be the negative impact of commenting out the lines in `clientConfiguration.cpp`? 
I understand that removing those lines will remove the need for the ec2 metadata check, but is there any negative impact of that for the code in version 1.8?

**Comment by KaibaLopez** [comment]
Not really, it just means that it won't do it for you.

**Comment by TimLand** [comment]
mark
